contrib/signet/README.md

```diff
@@ -17,3 +17,64 @@ Syntax: `getcoins.py [-h|--help] [-c|--cmd=<bitcoin-cli path>] [-f|--faucet=<fau
 If using the default network, invoking the script with no arguments should be sufficient under normal
 circumstances, but if multiple people are behind the same IP address, the faucet will by default only
 accept one claim per day. See `--password` above.
+
+miner
+=====
+
+To mine the first block in your custom chain, you can run:
+
+  cd src/
+  CLI="./bitcoin-cli -conf=mysignet.conf"
+  MINER="..contrib/signet/miner"
+  GRIND="./bitcoin-util grind"
+  ADDR=$($CLI -signet getnewaddress)
+  $MINER --cli="$CLI" generate --grind-cmd="$GRIND" --address="$ADDR" --set-block-time=-1
+
+This will mine a block with the current timestamp. If you want to backdate the chain, you can give a different timestamp to --set-block-time.
+
+You will then need to pick a difficulty target. Since signet chains are primarily protected by a signature rather than proof of work, there is no need to spend as much energy as possible mining, however you may wish to choose to spend more time than the absolute minimum. The calibrate subcommand can be used to pick a target, eg:
+
+  $MINER calibrate --grind-cmd="$GRIND"
+  nbits=1e00f403 for 25s average mining time
+
+It defaults to estimating an nbits value resulting in 25s average time to find a block, but the --seconds parameter can be used to pick a different target, or the --nbits parameter can be used to estimate how long it will take for a given difficulty.
+
+Using the --ongoing parameter will then cause the signet miner to create blocks indefinitely. It will pick the time between blocks so that difficulty is adjusted to match the provided --nbits value.
+
+  $MINER --cli="$CLI" generate --grind-cmd="$GRIND" --address="$ADDR" --nbits=1e00f403 --ongoing
+
+Other options
+-------------
+
+The --debug and --quiet options are available to control how noisy the signet miner's output is. Note that the --debug, --quiet and --cli parameters must all appear before the subcommand (generate, calibrate, etc) if used.
+
+Instead of specifying --ongoing, you can specify --max-blocks=N to mine N blocks and stop.
+
+Instead of using a single address, a ranged descriptor may be provided instead (via the --descriptor parameter), with the reward for the block at height H being sent to the H'th address generated from the descriptor.
+
+Instead of calculating a specific nbits value, --min-nbits can be specified instead, in which case the mininmum signet difficulty will be targeted.
+
+By default, the signet miner mines blocks at fixed intervals with minimal variation. If you want blocks to appear more randomly, as they do in mainnet, specify the --poisson option.
```

Comment:
How can i set this fixed interval? Slightly confused the difference between nbits and the interval itself. It sounds like interval is every 5 minutes try to mine a 25 seconds difficult block? What if I just want "blocks every 30 seconds"?

Reply:
> $GENERATE --cli="$CLI" "$@" generate --grind-cmd="$GRIND" --address="$ADDR" --nbits=$NBITS --multiminer=$MULTIMINER --ongoing --poisson

is what it looks like on my end. I think you want --ongoing --poisson.

Reply:
I don't think --poisson is the right thing, as --poisson just adds noise?

I guess the parameter is standby delay, which defaults to 0?

Where I think I'm confused is how setting nbits works because we use the next_block_delta to define how long to wait before mining (via sleeping), but then the grind command conceivably takes est_time(nbits) to mine? so it feels like we're double delaying based on nbits?

Reply:
You can't choose the interval; the difficulty will just get adjusted so the interval levels out at around 10 minutes. What you can target is a difficulty, so that once you reach that difficulty you keep mining a block every 10 minutes so that you stay at that difficulty. If you're not already at that difficulty, you instead choose a shorter/longer interval, so the difficulty is increased/decreased at the next retarget.

If you want to mine a bunch of blocks to get the chain started, then backdating via `--set-block-time` works. You could also use `--nbits=170f1372` (current mainnet nbits) which would make it try to raise the difficulty by mining quickly (though not faster than an average of 2m30).

There's no double delay, because everything operates on the block time field -- ie, last block had timestamp T, next block has timestamp T+INTERVAL. The standby-delay and backup-delay parameters mean you won't start mining the block until T+INTERVAL+DELAY, and you won't finish mining the block until T+INTERVAL+DELAY+GRINDTIME, but those delays don't affect the following block, which will have a timestamp of T+INTERVAL+INTERVAL2 (though again you might not start mining it for a further DELAY, and it won't be finished until GRINDTIME has also passed)

Reply:
Ahhh this makes sense I didn't realize normal difficulty adjustment rules are still used.

This text could go great in the Readme btw :)

Maybe incl a shell script template to backdate 100 blocks on start to seed a usable wallet?

Reply:
I think just `START=$(date +%s -d"1000 minutes ago")` and `--set-block-time=$START` would work for that -- even with poisson timing you should end up mining 100+ blocks before you catch up with real time. Maybe something like that should just be the default if you're mining the first block of a signet chain?